### DISCUSSION TITLE
Building a manylinux wheel fails with "Invalid python interpreter version"

### DISCUSSION THREAD

**Discussion by HerrMuellerluedenscheid** [PyO3/maturin#946] [Q&A] [answered]
I'm trying to build a python `manylinux` wheel from a `pyo3` application with `maturin`. The host machine is an Apple M1.
Unfortunatly, I'm kind of stuck with the following error message:

```
(.venv) [root@5a3121cdbf90 ormsgpack]# .venv/bin/maturin build --no-sdist --release --strip --manylinux 2014 -i /opt/python/cp39-cp39/bin/python --target x86_64-unknown-linux-gnu
📦 Including license file "LICENSE-APACHE"
📦 Including license file "LICENSE-MIT"
🔗 Found pyo3 bindings
💥 maturin failed
  Caused by: Invalid python interpreter version
```

I basically copy-pasted these [github action steps](https://github.com/aviramha/ormsgpack/blob/master/.github/workflows/release.yml#L23).

**Comment by messense** [comment, marked as answer]
You should not just copy-paste commands without understanding what it does and what it requires to work. That command was running in a manylinux docker container so it won't work for you on macOS without using Docker.

You can simply run `maturin build --no-sdist --release --strip --manylinux 2014 -i python3.9 --target x86_64-unknown-linux-gnu` instead and maturin will use a bundled Linux Python 3.9 configuration for you automatically when you're using maturin v0.12.15+.

**Reply by HerrMuellerluedenscheid** [reply]
Thanks for the swift reply and the lecturing.

> You should not just copy-paste commands without understanding what it does and what it requires to work. That command was running in a manylinux docker container so it won't work for you on macOS without using Docker.

I didn't mention this (sorry for leaving that detail out) but I actually ran this inside the `manylinux2014_aarch64` container to ensure that my local environment matches the one running inside the github action (`manylinux2014_x86_64` on `ubuntu`) as closely as possible.

> You can simply run maturin build --no-sdist --release --strip --manylinux 2014 -i python3.9 --target x86_64-unknown-linux-gnu instead and maturin will use a bundled Linux Python 3.9 configuration for you automatically when you're using maturin v0.12.15+.

That looks like it's the same command except for not specifying the python interpreter path, explicitly, right? Having that defined did not seem an issue inside the existing github action. This command fails with a linking error but I first need to have another look to understand what's going on there.  

But unfortunately the issue persists inside the container.

**Reply by messense** [reply]
Ahh, `manylinux2014_aarch64` container can only be used to build for `aarch64-unknown-linux-gnu` Rust target while you were trying to build for `x86_64-unknown-linux-gnu` target.

**Reply by messense** [reply]
So to build for `x86_64-unknown-linux-gnu` Rust target, use `manylinux2014_x86_64` is possible with Docker's qemu support, but it's gonna be around 6 times slower than native build.

**Reply by messense** [reply]
BTW, if you are going to build on GitHub Actions, I suggest you take a look at `maturin-actions`'s examples: https://github.com/messense/maturin-action#examples

**Reply by HerrMuellerluedenscheid** [reply]
Ah, alright! Thanks a lot that was really helpful!
Lesson learned for me: provide the *full* context when asking questions 😊
